webapp/api/github.py

```diff
@@ -0,0 +1,143 @@
+import os
+import flask
+from urllib.parse import urlencode
+from webapp import api
+
+GITHUB_CLIENT_ID = os.getenv("GITHUB_CLIENT_ID")
+GITHUB_CLIENT_SECRET = os.getenv("GITHUB_CLIENT_SECRET")
+
+
+class GithubAuthException(Exception):
+    pass
+
+
+class GithubApi:
+    """
+    Provides authentication for GitHub users. Helper methods are also provided
+    for checking organization access and getting user data from the Github API.
+    """
+
+    def __init__(self):
+        self.client_id = GITHUB_CLIENT_ID
+        self.client_secret = GITHUB_CLIENT_SECRET
+        self.base_url = "https://api.github.com/"
+        self.base_auth_url = "https://github.com/login/oauth/"
+        self.session = api.requests.Session()
+        self.session.headers["Accept"] = "application/json"
+
+    def authorize(self, scope=None, state=None):
+        """
+        Redirect to Github and request access to a user's data.
+        """
+        params = {
+            "client_id": self.client_id,
+        }
+
+        if scope:
+            params.update({"scope": scope})
+
+        if state:
+            params.update({"state": state})
+
+        auth_url = f"{self.base_auth_url}authorize?{urlencode(params)}"
+
+        return flask.redirect(auth_url)
+
+    def verify_auth(self, code):
+        """
+        Handles response after the redirect to Github. This response
+        determines if the user has allowed this application access. If we
+        were then we send a POST request for the access_key used to
+        authenticate requests to Github.
+        """
+        data = {
+            "code": code,
+            "client_id": self.client_id,
+            "client_secret": self.client_secret,
+        }
+
+        response = self.request(
+            "POST", self.base_auth_url, "access_token", data
+        )
+
+        try:
+            data = response.json()
+        except ValueError:
+            pass
+
+        if not response.ok or "error" in data or "access_token" not in data:
```

Comment:
Why are you testing for 3 different types of failure here? Are you sure that these are all cases we expect? Are you sure that all failure states wouldn't be covered by `not response.ok`?

I kinda feel you're still masking here. Error messages are for developers, not for the public. If you take 3 different things that can possibly go wrong and mask them with one generic message, you reduce the chance that a developer can see an error and immediately tell what went wrong. In this case they're gonna have to run the code themselves, throw a break point in above this statement and inspect the response themselves to see what actually happened.

Even if you wanted to catch all 3 of these possible cases and raise your own custom errors like `GithubAuthException`, I feel you should either create 1 custom exception per possible error state or modify the exception object in some way to indicate what exactly happened.

Ok I'll delve into a little theory here. There are basically 2 different things you could do when you think there might be an error:

1. Catch errors yourself and re-raise them
2. Let the code error naturally (e.g. in this case you could call [`response.raise_for_status()`](https://2.python-requests.org/en/master/api/#requests.Response.raise_for_status), which will pass through silently if `response.ok` is `True`, or raise an [`HTTPError`](https://2.python-requests.org/en/master/api/#requests.HTTPError) if it's `False`)

The argument for 1. is encapsulation. In certain theories, all your code layers should be self-contained and opaque. The client of your code shouldn't care *at all* what you're using inside your module. In this case you might, at some point in the future, want to swap out your usage of `request` for e.g. using `urllib` directly. If you let your layer raise `requests.exceptions.HTTPError` errors, then client code might come to rely on those errors in their code, and so if you stop using `requests` it could end up breaking their code. And so you should catch every possible error state and raise your own custom exceptions, because that way you have complete control over whether those exceptions change or not.

The argument for .2 (and against .1) is that, for those who do understand how the module works, it becomes much less clear what actually went wrong. `requests` is a pretty universal library, and so if any Python developer ever encounters a `requests.exceptions.HTTPError` they probably know what it means and what to do with it. Your custom errors are much harder because the dev has to read your documentation or your code to work out what they mean and how to use them.

Another argument for .2 in my opinion is about efficiently of development, [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it), [MVP](https://en.wikipedia.org/wiki/Minimum_viable_product) and [gold-plating](https://en.wikipedia.org/wiki/Gold_plating_(project_management)):

> gold plating is the phenomenon of working on a project or task past the point of diminishing returns

It is literally impossible to anticipate all the ways in which your code could fail, and so you could spend all month trying to make sure your layer of code is robust against every eventually, and you could end up creating hundreds of custom exceptions along the way, which then make it a huge task to learn the full interface of your module. And in most cases this will cover plenty of cases that might never come up.

For me, the two arguments for option 2. massively trump the benefits of option 1., especially when this code is really mostly for our benefit and not a professional package or community offering. My attitude to exception handling is - try to make sure the code will naturally fail in an obvious way (`response.raise_for_status()`), but beyond that don't bother over-engineering your error handling.

So, for example (one which came up in @carkod's work), if your website expects a `SECRET_KEY` environment variable, then use `os.environ["SECRET_KEY"]` to access it rather than `os.getenv("SECRET_KEY")`. This way, if the `SECRET_KEY` is missing, it will raise a `KeyError: 'SECRET_KEY'` error. This is easy, quick, and should provide the developer with a pretty obvious indication of what went wrong.

In conclusion, what I would do in your case is not both defining `GithubAuthException` and instead use `response.raise_for_status()` where appropriate. If you're sure that the API will sometimes return a `2xx` response but without the expected `access_token` property, then it might be a valid reason for writing and raising a custom exception. But if you're not absolutely sure that's an expected resposne, then simply access the property when you need it with `data["access_token"]` and it will raise an error naturally if ever that case comes up.

Does this make sense?

Reply:
Okay, I understand let's say that I use `raise_for_status()` and `data["access_token"]`. Then I have:
```python

    def verify_auth(self, code):
        ...
        response.raise_for_status()
        data = response.json()
        return data["access_token"]
```
So that means different kind of exceptions could happen, right? Then I need to notify the user that something went wrong during the authorization verification. So with one should I use? Because I need to have something like this, right?:
```python
@oauth.route("/github/auth/verify", methods=["GET"])
@login_required
def github_login_verify():
    ...
    try:
        auth_token = github.verify_auth(auth_code)
    except RequestException, KeyError:
        flask.flash(f"An error has occurred linking your GitHub account", "negative")
        return flask.redirect(flask.url_for("snapcraft.homepage"))
```

But if I do this where is the benefit for a developer when debugging?

Obviously I need to notify the user that something went wrong during the authorization or is it better to let the exceptions raise and show to the user a 500 error page?

Reply:
What I'm saying is, deliberately handle the states you expect. And write your code such that it will error clearly for states you don't expect (by using `dict["key"]` where we expect keys to exist, and `raise_for_status()` to raise errors for any unexpected statuses).

Your use of `except RequestException, KeyError` is, I'm afraid, still not doing this because it's catching a whole load of things that could happen (any non `2xx` status was returned from the API *or* a `2xx` status was returned but it didn't contain a token). It will then clobber any useful error information by just showing a "An error has occurred linking your GitHub account", when there could have been an important unexpected error that we needed to know about.

So I assume one of the cases you do expect is that the user might decline to grant permissions on the GitHub authorization page. In this case, what will happen in your code? I would expect that maybe the response will contain a `status_code` of `403 forbidden`. if this is the case, then check for that - e.g. `if response.status_code == 403:`.

Another possible case might be that the user doesn't log in at all on GitHub and returns to our site unauthenticated, in which case maybe the API returns a response with a `status_code` of `401`, so we should make sure we do something sensible there.

Now, if it is the case that that's what the API will return, then we could write it like this:

``` python3
# github.py
def verify_auth(self, code):
    ...
    response.raise_for_status()
    return response.json()["access_token"]

# views.py
from requests.exceptions import HTTPError

@oauth.route("/github/auth/verify", methods=["GET"])
@login_required
def github_login_verify():
    ...
    try:
        auth_token = github.verify_auth(auth_code)
    except HTTPError as error:
        if error.response.status_code == 403:
            flask.flash(f"We were forbidden from accessing your Github account", "negative")
            return flask.redirect(flask.url_for("snapcraft.homepage"))
        elif error.response.status_code == 401:
            flask.flash(f"*something about what happened here*", "negative")
            return flask.redirect(flask.url_for("snapcraft.homepage"))
        else:
            raise error  # Anything we didn't anticipate
```

The important point here is we're explicitly checking for the states we expect, and doing something, and for the states we don't expect (a non 401 or 403 status code) we are still raising an error. If the `else:` case is hit here, the user will see a `500` status page, which is what should happen because it was unexpected.

Do you see what I mean?

Don't just copy my solution because I have no idea if this is how the GitHub API behaves. Check how it actually does behave, either by reading docs or testing it out, and then code your solution explicitly to respond to how it's expected to behave.

Reply:
Although, are you sure that the homepage is where you want to go if you if there's an error authenticating with GitHub? Surely you would want to remain on the build page for your snap wouldn't you?

Reply:
> Now, if it is the case that that's what the API will return, then we could write it like this:

Another option here would use custom error messages to keep the `401` logic in the `verify_auth` method, which is arguable where it might belong:

``` python3
# github.py
from requests.exceptions import HTTPError


class Unauthorized(HTTPError):
    pass


class Forbidden(HTTPError):
    pass


def verify_auth(self, code):
    ...

    if response.status_code == 401:
        raise Unauthorized(response=response)
    elif response.status_code == 403:
        raise Forbidden(response=response)

    response.raise_for_status()

    return response.json()["access_token"]


# views.py
import github


@oauth.route("/github/auth/verify", methods=["GET"])
@login_required
def github_login_verify():
    ...
    try:
        auth_token = github.verify_auth(auth_code)
    except github.Forbidden:
            flask.flash(f"We were forbidden from accessing your Github account", "negative")
    except github.Unauthorized:
            flask.flash(f"We couldn't get authorization to GitHub", "negative")
```

This is probably a little neater by encapsulating more of the logic in the right place (`verify_auth`) and keeping the view code pretty obviously straightforward. But it still achieved the exact same logic: Explicitly check for the states we expect - 401 or 403 - and allow the code to error if any state we didn't expect occurs.